redact/redact.go

```diff
@@ -0,0 +1,59 @@
+package redact
+
+import (
+	"net/http"
+	"strings"
+)
+
+var SensitiveFields = []string{
```

Comment:
I like the idea of a better API around this.  What about

```go
type SensitiveFields []string

var defaultSensitiveFields = SensitiveFields{"authorization", "cookie", "token", "password", "secret"}

func IsSensitive(key string) bool {
	return defaultSensitiveFields.IsSensitive(key)
}

func (s SensitiveFields) IsSensitive(key string) bool {
	// The work
}
```

and similarly for `Map` and `Headers`?

Reply:
We want to expose the full list, since it can be reused for the Bugsnag configuration to redact the same keys (in the same manner) for Bugsnag reports.

Reply:
Bugsnag should either use this, including `IsSensitive` and the logic that captures, or not use this.  Not being able to get the list feels like a feature, not a bug, to me, since the list is not actually a list of what it blocks. If you just make it private, I'd be more than happy to PR the API changes I suggest.

Reply:
The Bugsnag go library needs a `[]string` for its ParamsFilter, https://github.com/bugsnag/bugsnag-go/blob/834d8fee64233ea79b0389ad7a5a438261788f66/configuration.go#L82-L86. If we want to reuse the list of sensitive fields both here in `redact` and in any Bugsnag setup, which I think we do, we need a way to get the `[]string`.

I agree that a mutable global slice isn't optimal, but I don't see any good way of doing it otherwise. We could of course extend your `SensitiveFields` with a `ToSlice() []string`. We'd need `Add` + `Remove` or `Set` methods too, if you want to extend the defaults, or remove something from them?

Reply:
I don't want people to modify the defaults, but instead create their own set:
```go
myRedact = redact.SensitiveFields{"private", "no-peeking"}
_ = myRedact.Map(whatever)
```

Reply:
I'm good with a `ToSlice()` that could `strings.ToLower`

Reply:
Hmm, let me take a step back and argue again for why I think a global list of sensitive substrings is good, and why mutating it is good.

Imagine you're building a service that does authorization through a header called `API-Key`. Or your building a http client that logs requests and responses, and you use it to an upstream service Foo, which does auth through the header `Foo-Key`. "foo" or "key" will never make it into the goose default list of sensitive substrings, but in your app, it's critical that wherever redaction is done, `Foo-Key` should be redacted.

Your app might have code that calls `redact.Map` directly, but goose will also call it internally (e.g. in the request logging middleware), and all those places in goose will now need to be configurable with a redact function, or a set of sensitive strings.

I think a global setting is simpler, with less room for error. It's similar to how [`logger.GlobalFields`](https://github.com/Shopify/goose/blob/c9f72561fff2d353e156a8d1efe9e96780e23c53/logger/logger.go#L15) work.

The API could be:

```go
var sensitiveSubstrings = []string{
	"authorization",
	"cookie",
	"token",
	"password",
	"secret",
}

func AddSensitiveSubstring(substrings ...string) {
	lowerSubstrings := make([]string, len(substrings))
	for i, s := range substrings {
		lowerSubstrings[i] = strings.ToLower(s)
	}
	sensitiveSubstrings = append(sensitiveSubstrings, lowerSubstrings...)
}

func SensitiveSubstrings() []string {
	r := make([]string, len(sensitiveSubstrings))
	copy(r, sensitiveSubstrings)
	return r
}

func IsSensitive(key string) bool {
	for _, sensitiveKey := range sensitiveSubstrings {
		if strings.Contains(strings.ToLower(key), strings.ToLower(sensitiveKey)) {
			return true
		}
	}
	return false
}
```

What do you think?

Reply:
I also believe that applications that use `redact` will only configure the sensitive substrings at boot (or through `func init() {}` functions, and that there is no immediate use-case to have it more dynamic than that.

Reply:
I’m really not a fan of relying on global state for this.

I understand that we use it for `logger.GlobalFields`, but it is explicitly _Global_ and is empty by default. It’s meant for things like configuring the hostname or the version of the app.

Here, I could see someone wanting to log non-redacted versions of `cookie` or `token`. 
I could also see someone wanting to disable this behaviour in dev or test environments.

I would much rather see an implementation like:
```go
func NewRequestMetricsMiddleware(opts *RequestMetricsOptions) func(http.Handler) http.Handler {
	return func RequestMetricsMiddleware(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		}
	}
}

var RequestMetricsMiddleware = NewRequestMetricsMiddleware(defaultOptions)
```

This doesn't break current users and allows future extensions, also without breaking users

Reply:
Hmm, I'd rather rename the field to `GlobalSensitiveSubstrings` then (but keep the default values), and remove `SensitiveSubstrings() []string`. An app that wants to log these in dev/test can then set this to an empty slice.

I agree it could be solved like with a `RequestMetricsOption`, but then all places that need redaction would need to implement such optional lists of sensitive substrings. Applications using them might forget that they need to provide them in those places too.

We also want applications to be able to build their sensitive fields as defaults + extras, so somehow accessing the defaults would be necessary

Reply:
My suggestion does use the defaults automatically if you use `RequestMetricsMiddleware` directly, it just allows you to skip the behaviour if you want.

And yes, I agree that accessing `RequestMetricsDefaultOptions()`, which would return a copy, would be a good idea.

`GlobalSensitiveSubstrings` is slightly better, I could go with that if you insist, but the non-global alternative is rather easy to achieve here (compared to the logger example)

Reply:
Regardless of the solution for the request logging middleware, I need the `GlobalSensitiveSubstrings` `[]string` to pass to the Bugsnag configuration. Because of that, and to ensure that Bugsnag uses the same set of substrings for redaction as goose and the rest of the application, I think a global list like this makes sense.

As I wrote above:

> I also believe that applications that use redact will only configure the sensitive substrings at boot (or through func init() {} functions, and that there is no immediate use-case to have it more dynamic than that.

If we find use-cases to have it more dynamic than "one set of redact rules per application", I'd be happy to change.

What do you think?

Reply:
Ok